analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/TestAnalysisComparer.cs

```diff
@@ -0,0 +1,132 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2021 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using Microsoft.CodeAnalysis;
+using SonarAnalyzer.Common;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.UnitTest.TestFramework
+{
+    public static class TestAnalysisComparer
+    {
+        public static StringBuilder Compare(Diagnostic[] diagnostics, Dictionary<string, IList<IIssueLocation>> expectedIssuesPerFile, string languageVersion)
+        {
+            var summary = new StringBuilder($"Comparing actual issues with expected:\nLanguage version: {languageVersion}.\n");
+            var actualIssues = diagnostics.ToIssueLocations();
+
+            foreach (var fileName in actualIssues.Keys.OrderBy(x => x))
+            {
+                summary.Append('\n').Append(fileName).Append('\n');
+
+                foreach (var actualIssue in actualIssues[fileName].OrderBy(x => x.LineNumber))
+                {
+                    var issue = expectedIssuesPerFile[fileName].FirstOrDefault(x => x.IsPrimary == actualIssue.IsPrimary && x.LineNumber == actualIssue.LineNumber);
+                    summary.Append(Compare(actualIssue, issue));
```

Comment:
To avoid the `new StringBuilder() inside. Most of the issues will not append anything
```suggestion
                    AppendComparison(summary, actualIssue, issue));
```
or an extension method (let me guess what you'll prefer 😄 )
```suggestion
                    summary.AppendComparison(actualIssue, issue));
```

Reply:
I like more the current version since:
- the number of parameters is small
- it self contained and it does not alter external state
- performance should not be a concern here

Reply:
It's +1 parameter from 2 to 3. Not really an issue.
Why not the extension method then? That one makes most sense.
We can do better in an easy way. There are a lot of diagnostics in our UTs.